src/librustdoc/html/markdown.rs

```diff
@@ -695,8 +695,23 @@ pub fn render(w: &mut fmt::Formatter,
             "".to_owned()
         } else {
             let bytes = unsafe { (*text).as_bytes() };
-            let s = str::from_utf8(bytes).unwrap();
-            collapse_whitespace(s)
+            if bytes.contains(&0) {
+                let mut full = Vec::with_capacity(bytes.len());
+                let mut added = false;
+                for slice in bytes.split(|&x| x == 0) {
+                    if !full.is_empty() {
+                        full.extend_from_slice(&[b'\\', b'0']);
+                        added = true;
+                    }
+                    full.extend_from_slice(slice);
+                }
+                if !added {
+                    full.extend_from_slice(&[b'\\', b'0']);
+                }
```

Comment:
This loop feels suspect. After a long enough read it *seems* okay, but maybe replacing it with a loop over indices and something like `full.splice(idx..(idx+1), b"\\0".iter().cloned())` would make it easier to read its intent.